File: HLRS/MPI/tasks/PY/Ch3/solutions/overtake.py
#!/usr/bin/env python3

################################################################# 
#                                                               # 
#  This file has been written as a sample solution to an        # 
#  exercise in a course given at the High Performance           # 
#  Computing Centre Stuttgart (HLRS).                           # 
#  The examples are based on the examples in the MPI course of  # 
#  the Edinburgh Parallel Computing Centre (EPCC).              # 
#  It is made freely available with the understanding that      # 
#  every copy of this file must include this header and that    # 
#  HLRS and EPCC take no responsibility for the use of the      # 
#  enclosed teaching material.                                  # 
#                                                               # 
#  Authors: Rolf Rabenseifner, Tobias Haas (HLRS)               # 
#                                                               # 
#  Contact: rabenseifner@hlrs.de                                # 
#                                                               # 
#  Purpose: A program to try MPI_Send and MPI_Recv.             # 
#                                                               # 
#  Contents: Python code, object send version (comm.send)       # 
#                                                               # 
#################################################################

from mpi4py import MPI

temp = None; mass = None
status = MPI.Status()

comm_world = MPI.COMM_WORLD
my_rank = comm_world.Get_rank()

if (my_rank == 0):
   temp=100.0
   mass=2.5
   comm_world.send(temp, dest=1, tag=17)
   comm_world.send(mass, dest=1, tag=18)

elif (my_rank == 1):
   mass = comm_world.recv(source=0, tag=18, status=status); # status optional
   print(f"I am process {my_rank} and received mass={mass:3.1f}")
   temp = comm_world.recv(source=0, tag=17);
   print(f"I am process {my_rank} and received temp={temp:3.0f}")

File: HLRS/MPI/tasks/C/Ch3/solutions/overtake.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the High Performance           *
 * Computing Centre Stuttgart (HLRS).                           *
 * The examples are based on the examples in the MPI course of  *
 * the Edinburgh Parallel Computing Centre (EPCC).              *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * HLRS and EPCC take no responsibility for the use of the      *
 * enclosed teaching material.                                  *
 *                                                              *
 * Authors: Rolf Rabenseifner (HLRS)                            *
 *                                                              *
 * Contact: rabenseifner@hlrs.de                                *
 *                                                              *
 * Purpose: A program to try MPI_Send and MPI_Recv.             *
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/

#include <stdio.h>
#include <mpi.h>

int main(int argc, char *argv[])
{
  int my_rank;
  float temp, mass;
  MPI_Status status;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    if (my_rank == 0)
    {
      temp=100;  mass=2.5;
      MPI_Send(&temp, 1, MPI_FLOAT, 1, 17, MPI_COMM_WORLD);
      MPI_Send(&mass, 1, MPI_FLOAT, 1, 18, MPI_COMM_WORLD);
    }
    else if (my_rank == 1)
    {
      MPI_Recv(&mass, 1, MPI_FLOAT, 0, 18, MPI_COMM_WORLD, &status);
      printf("I am process %i and received mass=%3.1f \n", my_rank, mass);
      MPI_Recv(&temp, 1, MPI_FLOAT, 0, 17, MPI_COMM_WORLD, &status);
      printf("I am process %i and received temp=%3.0f \n", my_rank, temp);
    }

  MPI_Finalize();
}
